[mypyc] Support attributes that override properties

JukkaL opened the pull request:
Code like this is now supported by mypyc:
```
class B:
    @property
    def x(self) -> int:
        return 0

class C(B):
    x: int = 0 # Attribute overrides a property
```

The implementation generates implicit getter/setter methods for attributes as needed and puts them in the vtable. I had to change both the irbuild "prepare" pass (where we generate declarations), irbuild main pass (where we generate implicit accessor IRs), and codegen (where implicit properties aren't visible externally to CPython).

Also fix a minor mypy bug related to overriding properties and multiple inheritance that I encountered. 

This doesn't handle glue methods yet.

JukkaL commented on mypyc/irbuild/prepare.py:
```diff
@@ -223,8 +236,68 @@ def prepare_class_def(
 
-    # We sort the table for determinism here on Python 3.5
-    for name, node in sorted(info.names.items()):
+    # Check for subclassing from builtin types
```
I had to restructure this to do things in a different order. Most of the logic hasn't changed. I also extracted some functionality to separate methods since the function was already quite big.

thomkeh commented:
Is the eventual goal to allow this for mypyc? (Which throws an exception in CPython)
```python
class B:
    @property
    def x(self) -> int:
        return 0

class C(B):
    x: int
    def __init__(self) -> None:
        self.x = 4
```

JukkaL commented:
Generally the goal is to reject things that don't work in CPython. Updated the example to something that actually works in CPython:
```py
class B:
    @property
    def x(self) -> int:
        return 0

class C(B):
    x: int = 0 # Attribute overrides a property
```
Added #14413 to track the mypy false negative when there is no initializer.

I'll also need to remove cases that aren't supported by CPython from the tests in added in this PR.

AlexWaygood commented:
I always thought that the idiomatic way to override a read-only property with a writable attribute was supposed to be this, but mypy unfortunately emits a false positive for this snippet:

```python
class A:
    _x: int
    @property
    def x(self) -> int:
        return self._x

class B(A):
    @A.x.setter  # mypy complains with 'error: "Callable[[A], int]" has no attribute "setter"  [attr-defined]'
    def x(self, val: int) -> None:
        self._x = val

b = B()
b.x = 42
b.x
```

JukkaL commented:
> I always thought that the idiomatic way to override a read-only property with a writable attribute was supposed to be this ...

This idiom is almost nonexistent in the codebases I work on, though mypy not supporting it may be the reason why developers use something different. Still, it would be nice to support it. I focused on the particular use case addressed in this PR since I have a follow-up PR that needs this.
